File: CryptoCurrencyChampion/app/src/main/java/com/ryan/cryptocurrencychampion/UI/MainActivity.java
package com.ryan.cryptocurrencychampion.UI;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Bundle;
import android.provider.Settings;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.Toast;
import com.ryan.cryptocurrencychampion.R;
import com.ryan.cryptocurrencychampion.RecyclerView.RecyclerViewAdapter;
import com.ryan.cryptocurrencychampion.RetroFit.AllCryptoCurrencyInfo;
import com.ryan.cryptocurrencychampion.RetroFit.CoinMarketCapAPI;
import com.ryan.cryptocurrencychampion.RetroFit.LatestCurrencyPojo.CryptoCurrency;
import com.ryan.cryptocurrencychampion.RetroFit.MetadataPojo.Metadata;
import com.ryan.cryptocurrencychampion.RetroFit.RetroFitInstance;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.functions.BiFunction;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;

public class MainActivity extends AppCompatActivity {

    private CoinMarketCapAPI handler;
    private RecyclerView recyclerView;
    private RecyclerViewAdapter recyclerViewAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        if(!isConnected())
        {
            showDialog();
        }
        recyclerView = findViewById(R.id.recyclerview);
        recyclerView.setHasFixedSize(true);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        callEndpoints("market_cap");
    }

    private void callEndpoints(String s) {

        if(!isConnected())
        {
            showDialog();
        }

        else
        {
            handler = RetroFitInstance.getRetrofitInstance().create(CoinMarketCapAPI.class);
            Observable<CryptoCurrency> observable = handler.getLatest(s);
            observable
                    .flatMap(new Function<CryptoCurrency, Observable<Metadata>>() {
                        @Override
                        public Observable<Metadata> apply(CryptoCurrency cryptoCurrency) throws Exception {
                            return handler.getMetaData(cryptoCurrency.getIdQuery());
                        }
                    }, new BiFunction<CryptoCurrency, Metadata, AllCryptoCurrencyInfo>() {
                        @Override
                        public AllCryptoCurrencyInfo apply(CryptoCurrency cryptoCurrency, Metadata metadata) throws Exception {
                            return new AllCryptoCurrencyInfo(metadata, cryptoCurrency);
                        }
                    })
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(this::handleResults, this::handleError);
        }
    }


    private void handleResults(AllCryptoCurrencyInfo allCryptoCurrencyInfo) {
        if (allCryptoCurrencyInfo.getCryptoCurrency().getData() != null
                && allCryptoCurrencyInfo.getCryptoCurrency().getData().size() != 0) {
            if (recyclerViewAdapter == null)
            {
                recyclerViewAdapter = new RecyclerViewAdapter(allCryptoCurrencyInfo, this);
                recyclerView.setAdapter(recyclerViewAdapter);
            }
            else
            {
                recyclerViewAdapter.update(allCryptoCurrencyInfo);
            }
        } else {
            Toast.makeText(this, "NO RESULTS FOUND",
                    Toast.LENGTH_LONG).show();
        }
    }

    private void handleError(Throwable t) {

        Toast.makeText(this, "ERROR IN FETCHING API RESPONSE. Try again",
                Toast.LENGTH_LONG).show();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.main_activity_menu, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        if(id==R.id.sort)
        {
            showSortDialog();
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    private void showSortDialog() {
        String [] options = {"Market Cap", "Name", "Price", "Date Added", "24 Hour Volume"};
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle("Sort By");
        builder.setIcon(R.drawable.ic_action_sort);
        builder.setItems(options, new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialogInterface, int i) {
                switch (i)
                {
                    case 0:
                        callEndpoints("market_cap");
                        break;
                    case 1:
                        callEndpoints("name");
                        break;
                    case 2:
                        callEndpoints("price");
                        break;
                    case 3:
                        callEndpoints("date_added");
                        break;
                    case 4:
                        callEndpoints("volume_24h");
                        break;
                }
            }
        });
        builder.create().show();
    }

    private boolean isConnected() {
        boolean isConnectedWifi = false;
        boolean isConnectedMobile = false;

        ConnectivityManager cm = (ConnectivityManager) getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo[] netInfo = cm.getAllNetworkInfo();
        for (NetworkInfo ni : netInfo) {
            if (ni.getTypeName().equalsIgnoreCase("WIFI"))
                if (ni.isConnected())
                    isConnectedWifi = true;
            if (ni.getTypeName().equalsIgnoreCase("MOBILE"))
                if (ni.isConnected())
                    isConnectedMobile = true;
        }
        return isConnectedWifi || isConnectedMobile;
    }

    private void showDialog()
    {
        AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setMessage("Connect to wifi or quit")
                .setCancelable(false)
                .setPositiveButton("Connect to WIFI", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface dialog, int id) {
                        startActivity(new Intent(Settings.ACTION_WIFI_SETTINGS));
                    }
                })
                .setNegativeButton("Quit", null);
        AlertDialog alert = builder.create();
        alert.show();
    }
}

File: app/src/main/java/com/ryan/redditbrowser/UI/MainActivity.kt
package com.ryan.redditbrowser.UI

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkInfo
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.View
import androidx.appcompat.widget.SearchView
import androidx.core.view.isVisible
import androidx.recyclerview.widget.DividerItemDecoration
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.ryan.redditbrowser.R
import com.ryan.redditbrowser.Retrofit.Data.RedditResponse
import com.ryan.redditbrowser.Retrofit.RedditAPI
import com.ryan.redditbrowser.UI.Recyclerview.RecyclerViewAdapter
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.observers.DisposableSingleObserver
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private val redditAPI by lazy { RedditAPI.create() }

    private lateinit var searchView: SearchView

    private var dataAdapter: RecyclerViewAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        val layoutManager: RecyclerView.LayoutManager = LinearLayoutManager(this)
        recyclerview.layoutManager = layoutManager
        recyclerview.visibility = View.INVISIBLE
        constraintLayout.setBackgroundResource(R.drawable.main_screen)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.menu, menu)
        val searchItem = menu!!.findItem(R.id.action_search)
        searchView = searchItem.actionView as SearchView
        searchView.setSubmitButtonEnabled(true)
        searchView.setIconifiedByDefault(false);
        searchView
        searchView.setQueryHint("Search for a subreddit")
        searchView.setOnQueryTextListener(object : SearchView.OnQueryTextListener {
            override fun onQueryTextChange(newText: String): Boolean {
                return true
            }

            override fun onQueryTextSubmit(query: String): Boolean {
                search(query)
                searchView.clearFocus()
                return true
            }
        })
        return super.onCreateOptionsMenu(menu)
    }

    private fun search(searchString: String) {
        if (!isConnected()) {
            if (recyclerview.isVisible) {
                recyclerview.visibility = View.INVISIBLE
            }
            constraintLayout.setBackgroundResource(R.drawable.nointernetsaccess)
        } else {
            var disposableSingleObserver = object : DisposableSingleObserver<RedditResponse>() {
                override fun onSuccess(t: RedditResponse) {
                    if (!t.data.children.isEmpty()) {
                        recyclerview.visibility = View.VISIBLE
                        constraintLayout.setBackgroundResource(0)
                        if (dataAdapter == null) {
                            dataAdapter = RecyclerViewAdapter(t.data.children, this@MainActivity)
                            recyclerview.adapter = dataAdapter
                        } else {
                            dataAdapter!!.swap(t.data.children)
                        }
                    } else {
                        if (recyclerview.isVisible) {
                            recyclerview.visibility = View.INVISIBLE
                        }
                        constraintLayout.setBackgroundResource(R.drawable.nosubbreddits)
                    }
                }

                override fun onError(e: Throwable) {
                    if (recyclerview.isVisible) {
                        recyclerview.visibility = View.INVISIBLE
                    }
                    constraintLayout.setBackgroundResource(R.drawable.nosubbreddits)
                }

            }
            var searchQuery: Single<RedditResponse> = redditAPI.getData(searchString)
            searchQuery.subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(disposableSingleObserver)
        }
    }

    fun isConnected(): Boolean {
        val cm = this.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val activeNetwork: NetworkInfo? = cm.activeNetworkInfo
        val isConnected: Boolean = activeNetwork?.isConnectedOrConnecting == true
        return isConnected
    }
}
